[14.0] shopfloor: add get work option on location content transfer

TDu opened the pull request:
ref.: rau-75

TDu committed:
shopfloor: add get work option

TDu committed:
shopfloor: add get work option on location content transfer

TDu committed:
shopfloor: add location content transfer diagram

simahawk commented:
pls fix tests https://github.com/OCA/wms/actions/runs/3524766119/jobs/5910605264#step:8:1809

simahawk commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -51,9 +51,32 @@ class LocationContentTransfer(Component):
+            return self._response(
+                next_state="get_work", data={}, message=message, popup=popup
+            )
+        # TODO change the next_state to `scan_location`
```
TODO?

TDu replied:
This is fix by the first fixup

simahawk commented on shopfloor/services/location_content_transfer.py:
```diff
@@ -258,6 +280,54 @@ def _unreserve_other_lines(self, location, move_lines):
+            order="create_date",
+            limit=1,
+        )
+        return fields.first(next_picking.move_line_ids).location_id
```
don't we filter on qty_done not full?

mmequignon replied:
as soon `shopfloor_reception` is merged, you'll be able to add `('progress', '<', 100)` in the domain.

TDu replied:
> don't we filter on qty_done not full?

Yes, added.

simahawk reviewed (commented):
LG overall

simahawk reviewed (approved):
Pls squash. Still one comment w/o answer.

mmequignon commented on shopfloor/actions/stock.py:
```diff
@@ -35,6 +35,21 @@ def mark_move_line_as_picked(
 
+    def unmark_move_line_as_picked(self, move_lines):
+        """Reverse the change from `mark_move_line_as_picked`."""
+        move_lines.write(
+            {
+                "shopfloor_user_id": False,
+                "qty_done": 0,
+            }
+        )
+        move_lines.move_id.picking_id.write(
+            {
+                "user_id": False,
+                "printed": False,
+            }
+        )
+
```
as it is meant to revert changes done by `mark_move_line_as_picked`, shouldn't we remove `result_package_id` as well?

TDu replied:
Yeah, done.

mmequignon reviewed (approved):
LG

simahawk commented:
@TDu ping :)

jbaudoux commented on shopfloor/data/shopfloor_scenario_data.xml:
```diff
@@ -53,7 +53,8 @@
     "allow_unreserve_other_moves": true,
-    "allow_ignore_no_putaway_available": true
+    "allow_ignore_no_putaway_available": true,
+    "allow_get_work": false
```
```suggestion
    "allow_get_work": true
```

TDu replied:
ok

jbaudoux commented on shopfloor_base/models/shopfloor_scenario.py:
```diff
@@ -77,4 +77,4 @@ def _normalize_key(name):
 
     def has_option(self, key):
-        return self.options.get(key, False)
+        return key in self.options.keys()
```
Don't change this, set new option to True

TDu replied:
ok

TDu committed:
shopfloor: imp location content transfer

This part is removed because it is done by `mark_move_line_as_picked`
further down

jbaudoux reviewed (approved):
Tested and works fine

TDu commented:
I just squashed the fixups

sebalix commented:
LG, thanks for the documentation!
/ocabot merge minor

jbaudoux commented on shopfloor/actions/stock.py:
```diff
@@ -35,6 +35,22 @@ def mark_move_line_as_picked(
+                "result_package_id": False,
+            }
+        )
+        move_lines.move_id.picking_id.write(
```
1/ move lines could be not linked to moves, you should take the picking_id from the move_lines directly
2/ you should check that there are no other move lines in the picking that still have a shopfloor_user_id, otherwise you should extract current move lines in a split order with `_extract_in_split_order`. This will ensure that we don't end-up with a picking without user_id still having move lines with a shopfloor_user_id
